shared/middle/sicp.md

```diff
@@ -0,0 +1,214 @@
+    * What is the propagation of constraints?
+    * How does constraint networks work?
+    * How are assignments (or mutability) used in implementation of constraint systems?
+    * What is the expression-oriented style? What are advantages of this style comparing to imperative style?
+
+  * [3.4 Concurrency: Time Is of the Essence](http://mitpress.mit.edu/sites/default/files/sicp/full-text/book/book-Z-H-23.html#%_sec_3.4)
+
+    * What does concurrency mean?
+    * What is the shared state? How it can be done? What problems can appear because of shared state?
+    * Explain all possible interleaving the order of events for two processes when first one executes the expression: `balance += 10;` and the second one: `balance *= 2;`?
+    * Will be the problems with concurrency same if we had no assignment in the language (only immutable data)?
+    * Why assignments in the language forces programmer to admit time in the modeling?
+    * What does serialization mean and what is its purpose?
+    * Why serialization doesn't help for a such operation like swapping the balances in two bank accounts?
+    * What is the mutex? How does it help with problems arisen because of concurrency?
+    * How does mutex can be implemented?
+    * What operations in mutex implementation should be done atomically? Why?
+    * Which situations are called deadlock? What are the ways to avoid deadlocks?
+    * Why does communication between processes matter for shared state or ordering of events?
+
+  * [3.5 Streams](http://mitpress.mit.edu/sites/default/files/sicp/full-text/book/book-Z-H-24.html#%_sec_3.5)
+
+    * What the difference between lists and streams? What is happening in both construction and selection time for both lists and streams?
+    * Why cons-stream and delay should be special forms of languages? Should `force` be such procedure?
+    * Why should all list-manipulating procedures be rewritten for streams? What actually should be changed?
+    * What's the difference between methods call-by-name and call-by-need?
+    * Explain the behavior of the programs from exercises 3.51 and 3.52 (you don't need to complete exercises if you can fully explain the behavior).
+    * Why streams can be infinite? Can we make our program stuck when working with infinite streams?
+    * What's the order of growth of the computing of fibonacci numbers from stream declared implicitly both using call-by-name and call-by-need strategies?
+    * What's the order of growth of the stream of all positive integers with no prime factors other than 2, 3, or 5 declared implicitly?
+    * What's the technique called "sequence accelerator"?
+    * How can we can accelerate the accelerated sequence?
+    * How can we merge infinite sequences? Why simple appending doesn't fit here?
+    * How does delayed evaluation help to create circular dependencies?
+    * What should be changed in the language, so it become normal-order evaluated? What are pros and cons of this approach?
```

Comment:
What should be changed in the language to become normal-order evaluated?

Reply:
so it becomes вроде получше?

Reply:
В таких предложениях часто ставят that, so that, in order that, т.е. _What should be changed in the language so that it become normal-order evaluated?_.
В придаточном цели используют subjunctive I - форму глагола без -s, так что become без -s допускается. Эта форма в разговорной речи почти не используется, но в статьях попадается часто.

Reply:
А получилось в итоге хорошо?

Reply:
Ну вот мне кажется, что that недостает. Я не нашел варианта для сложного предложения с разными подлежащими без that или lest, оно везде есть. Можно сформулировать с одним подлежащим: _What should be changed in the language to make it normal-order evaluated?_.

https://www.englishgrammar.org/adverb-clauses-purpose/

И я не уверен, что можно написать that it becomes - видел that it become (без -s, книжный/старый вариант) либо that it might become (более современный). Возможно, так говорят, не знаю.

Тут речь идет о конкретном языке или о любом? Если о любом, то the language -> a language.

Reply:
Тут речь идет о scheme, ну или в общем о языке, который является applicative order evaluated

Reply:
Тоже проверил этот вопрос в grammarly, в другом месте была ошибка